vllm/v1/engine/core.py

```diff
@@ -367,44 +368,70 @@ def __init__(
+                             executor_fail_callback)
+
+        self.step_fn = (self.step if self.batch_queue is None else
+                        self.step_with_batch_queue)
+
+        # Background Threads and Queues for IO. These enable us to
+        # overlap ZMQ socket IO with GPU since they release the GIL,
+        # and to overlap some serialization/deserialization with the
+        # model forward pass.
+        # Threads handle Socket <-> Queues and core_busy_loop uses Queue.
+        threading.Thread(target=self.process_input_sockets,
+                         args=(self.addresses.inputs,
+                               self.addresses.coordinator_input,
+                               self.identity),
+                         daemon=True).start()
+        self.output_thread = threading.Thread(
+            target=self.process_output_sockets,
+            args=(self.addresses.outputs, self.addresses.coordinator_output,
+                  self.engine_index),
+            daemon=True)
+        self.output_thread.start()
+
+    @contextmanager
+    def _perform_handshake(
+            self, handshake_address: str, identity: bytes, on_head_node: bool,
+            vllm_config: VllmConfig
+    ) -> Generator[EngineZmqAddresses, None, None]:
         input_ctx = zmq.Context()
-        identity = engine_index.to_bytes(length=2, byteorder="little")
         with make_zmq_socket(input_ctx,
                              handshake_address,
                              zmq.DEALER,
                              identity=identity,
                              linger=5000,
                              bind=False) as handshake_socket:
-
-            # Register engine with front-end.
+            # Register engine with front-end
             addresses = self.startup_handshake(handshake_socket, on_head_node,
                                                vllm_config.parallel_config)
-            self.client_count = len(addresses.outputs)
 
-            # Update config which may have changed from the handshake.
+            # Update config which may have changed from the handshake
             vllm_config.__post_init__()
 
-            # Set up data parallel environment.
-            self.has_coordinator = addresses.coordinator_output is not None
-            self._init_data_parallel(vllm_config)
-
-            # Initialize engine core and model.
-            super().__init__(vllm_config, executor_class, log_stats,
-                             executor_fail_callback)
-
-            self.engine_index = engine_index
-            self.step_fn = (self.step if self.batch_queue is None else
-                            self.step_with_batch_queue)
-            self.engines_running = False
-            self.last_counts = (0, 0)
+            yield addresses
 
-            # Send ready message.
+            # Send ready message
             num_gpu_blocks = vllm_config.cache_config.num_gpu_blocks
             handshake_socket.send(
```

Comment:
This will run even if the startup raises an exception ... I don't think we want that. We should at least return "FAILED" status in this case.

Reply:
OK sorry I had bad information on this!